### PR TITLE
Clarify error message for bad keyword arguments.

### PR DESCRIPTION
**Pull request by anntzer**
`plot([], [], foo=42)` previously emitted
```
'Line2D' object has no property 'foo'
```
which refers to the Matplotlib-specific concept of "properties".  It now
instead emits
```
Line2D.set() got an unexpected keyword argument 'foo'
```
which is modeled after the standard error message for unknown keyword
arguments.

(To maximize backcompat, the implementation goes through a new
_internal_update, which does *not* error when the same prop is passed
under different aliases.  This could be changed later, but is not the
goal of this PR.)

(This was triggered by https://discourse.matplotlib.org/t/line2d-object-has-no-property-line-error/22585)

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/artist.py [hunk 1]
```diff
@@ -1161,12 +1135,53 @@ def properties(self):
...
+
     def set(self, **kwargs):
         # docstring and signature are auto-generated via
         # Artist._update_set_signature_and_docstring() at the end of the
         # module.
-        kwargs = cbook.normalize_kwargs(kwargs, self)
-        return self.update(kwargs)
+        return self._update_props("set", cbook.normalize_kwargs(kwargs, self))
+
+    def _internal_update(self, **kwargs):
```
Comment (QuLogic) [inline lib/matplotlib/artist.py:1167]: I don't think there's any caller that passes individual args, so is there a need for the `**`?

File: lib/matplotlib/artist.py [hunk 2]
```diff
@@ -1161,12 +1135,53 @@ def properties(self):
         """Return a dictionary of all the properties of the artist."""
         return ArtistInspector(self).properties()
 
+    def _update_props(self, caller, props):
+        """Helper for `set` and `update`."""
+        errorfmt = {  # Often called & only internal, so use plain dict access.
+            "update": "{.__name__!r} object has no property {!r}",
+            "set": "{.__name__}.set() got an unexpected keyword argument {!r}",
+        }[caller]
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/artist.py:1143]: yes, that's better.

File: lib/matplotlib/artist.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/artist.py]: sure.

File: lib/matplotlib/artist.py [hunk 3]
```diff
@@ -1161,12 +1135,57 @@ def properties(self):
         """Return a dictionary of all the properties of the artist."""
         return ArtistInspector(self).properties()
 
+    def _update_props(self, props, errorfmt):
+        """
+        Helper for `set` and `update`.
```
Comment (timhoffm) [inline lib/matplotlib/artist.py:1140]: Clarification:
```suggestion
        Helper for ``Artist.set`` and ``Artist.update``.
```

File: lib/matplotlib/artist.py [hunk 4]
```diff
@@ -1161,12 +1135,57 @@ def properties(self):
...
+
+        *errorfmt* is used to generate error messages for invalid property
+        names; it get formatted with ``type(self)`` and the property name.
```
Comment (timhoffm) [inline lib/matplotlib/artist.py:1143]: I suggest to use keywords instead of positional arguments because definition and resolution are quite far apart. Naming the parameters helps to understand what is what and reduces the risk that somebody will mess this up later.

File: lib/matplotlib/collections.py [hunk 5]
```diff
@@ -1132,7 +1132,7 @@ def legend_elements(self, prop="colors", num="auto",
 
         kw = dict(markeredgewidth=self.get_linewidths()[0],
                   alpha=self.get_alpha())
-        kw.update(kwargs)
+        kw.update(**kwargs)
```
Comment (timhoffm) [inline lib/matplotlib/collections.py:1135]: Why this change? Both should work, but the original one is easier to read.

File: lib/matplotlib/artist.py (hunk 4 above)
Comment (anntzer) [reply lib/matplotlib/artist.py:1143]: sure

File: lib/matplotlib/collections.py (hunk 5 above)
Comment (anntzer) [reply lib/matplotlib/collections.py:1135]: sorry, reverted, that was just a bad search/replace.

File: lib/matplotlib/artist.py (hunk 4 above)
Comment (timhoffm) [reply lib/matplotlib/artist.py:1143]: Your comment sounds like you want to do this, but the code has not yet changed.

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/artist.py (hunk 4 above)
Comment (anntzer) [reply lib/matplotlib/artist.py:1143]: It has? errfmt is now kwonly (I'd argue there's little risk of misunderstanding what props is given the context).  Or did I misunderstand your suggestion?

Comment (timhoffm) [reply lib/matplotlib/artist.py:1143]: Oh, sorry. I was thinking about the format string, but didn't communicate that propertly; i.e.
```
errfmt.format(class_=type(self), prop_name=k)
```
and
 ```
"{class_.__name__}.set() got an unexpected keyword argument {prop_name!r}"
```

**Review by timhoffm — APPROVED** [review]

**Review by QuLogic — APPROVED** [review]
